engine/table/src/main/java/io/deephaven/engine/table/impl/BlinkTableTools.java

```diff
@@ -183,4 +108,133 @@ public static boolean isBlink(Table table) {
+
+        private BlinkToAppendOnlyOperation(
+                @NotNull final QueryTable parent,
+                final long sizeLimit,
+                @Nullable final Object memoKey) {
+            this.parent = parent;
+            this.sizeLimit = sizeLimit;
+            this.memoKey = memoKey;
+
+            this.sourceColumns = new ColumnSource<?>[parent.numColumns()];
+            this.destColumns = new WritableColumnSource<?>[parent.numColumns()];
+        }
+
+        @Override
+        public String getDescription() {
+            return "BlinkTableTools.blinkToAppendOnly(" + (sizeLimit == Long.MAX_VALUE ? "" : sizeLimit) + ")";
+        }
+
+        @Override
+        public String getLogPrefix() {
+            return "BlinkTableTools.blinkToAppendOnly";
+        }
+
+        @Override
+        public MemoizedOperationKey getMemoizedOperationKey() {
+            return memoKey == null ? null : MemoizedOperationKey.blinkToAppendOnly(sizeLimit, memoKey);
+        }
+
+        @Override
+        public Result<QueryTable> initialize(boolean usePrev, long beforeClock) {
+            final Map<String, WritableColumnSource<?>> resultColumns = new LinkedHashMap<>();
+            final Map<String, ? extends ColumnSource<?>> columnSourceMap = parent.getColumnSourceMap();
+
+            // note that we do not need to enable prev tracking for an add-only table
+            int colIdx = 0;
+            for (Map.Entry<String, ? extends ColumnSource<?>> nameColumnSourceEntry : columnSourceMap
+                    .entrySet()) {
+                final ColumnSource<?> existingColumn = nameColumnSourceEntry.getValue();
+                final WritableColumnSource<?> newColumn = ArrayBackedColumnSource.getMemoryColumnSource(
+                        0, existingColumn.getType(), existingColumn.getComponentType());
+                resultColumns.put(nameColumnSourceEntry.getKey(), newColumn);
+
+                // read and write primitives whenever possible
+                sourceColumns[colIdx] = ReinterpretUtils.maybeConvertToPrimitive(existingColumn);
+                destColumns[colIdx++] = ReinterpretUtils.maybeConvertToWritablePrimitive(newColumn);
+            }
+
+            final TrackingWritableRowSet rowSet = RowSetFactory.empty().toTracking();
+            resultTable = new QueryTable(rowSet, resultColumns);
+            resultTable.setAttribute(Table.ADD_ONLY_TABLE_ATTRIBUTE, true);
+            resultTable.setAttribute(Table.APPEND_ONLY_TABLE_ATTRIBUTE, true);
+            resultTable.setFlat();
+
+            // statistically a blink table is likely to be empty; but let's copy whatever we can
+            appendRows(usePrev ? parent.getRowSet().prev() : parent.getRowSet(), usePrev).close();
+            rowSet.initializePreviousValue();
+            Assert.leq(rowSet.size(), "rowSet.size()", sizeLimit, "sizeLimit");
+
+            if (resultTable.size() < sizeLimit) {
+                resultTable.setRefreshing(true);
```

Comment:
It should never be ambiguous whether we will get a refreshing result. This should not be conditional.